arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+    if (it.has_value()) {
+      delayedFuture(it->_executionDelay).thenFinal([it = it, logMetrics](auto&&) mutable {
+        // _follower is an alias of _parentLog
+        TRI_ASSERT(!it->_parentLog.owner_before(it->_follower) &&
+                   !it->_follower.owner_before(it->_parentLog));
+        // Let's check whether our log is still there, and abort otherwise
+        auto logLeader = it->_parentLog.lock();
+        auto follower = it->_follower.lock();
+        // as mentioned before, both are owning the same object, thus:
+        TRI_ASSERT((follower == nullptr) == (logLeader == nullptr));
+        if (logLeader == nullptr) {
+          LOG_TOPIC("de312", TRACE, Logger::REPLICATION2)
+              << "parent log already gone, not sending any more "
+                 "AppendEntryRequests";
+          return;
+        }
+
+        auto [request, lastIndex] =
+            logLeader->_guardedLeaderData.doUnderLock([&](auto const& self) {
+              auto lastAvailableIndex = self._inMemoryLog.getLastTermIndexPair();
+              LOG_CTX("71801", TRACE, follower->logContext)
+                  << "last acked index = " << follower->lastAckedEntry
+                  << ", current index = " << lastAvailableIndex
+                  << ", last acked commit index = " << follower->lastAckedCommitIndex
+                  << ", current commit index = " << self._commitIndex;
+              // We can only get here if there is some new information for this follower
+              TRI_ASSERT(follower->lastAckedEntry.index != lastAvailableIndex.index ||
+                         self._commitIndex != follower->lastAckedCommitIndex);
+
+              return self.createAppendEntriesRequest(*follower, lastAvailableIndex);
+            });
+
+        auto messageId = request.messageId;
+        LOG_CTX("1b0ec", TRACE, follower->logContext)
+            << "sending append entries, messageId = " << messageId;
+        auto startTime = std::chrono::steady_clock::now();
+        // Capture a weak pointer `parentLog` that will be locked
+        // when the request returns. If the locking is successful
+        // we are still in the same term.
+        follower->_impl->appendEntries(std::move(request))
+            .thenFinal([weakParentLog = it->_parentLog, weakFollower = it->_follower,
+                        lastIndex = lastIndex, currentCommitIndex = request.leaderCommit,
+                        currentTerm = logLeader->_currentTerm,
+                        messageId = messageId, startTime, logMetrics = logMetrics](
+                           futures::Try<AppendEntriesResult>&& res) noexcept {
+              // This has to remain noexcept, because the code below is not exception safe
+              auto const endTime = std::chrono::steady_clock::now();
+
+              if (auto self = weakParentLog.lock()) {
+                // If we successfully acquired parentLog, this cannot fail.
+                auto follower = weakFollower.lock();
```

Comment:
Should we add an assertion here that `follower` is not a nullptr?

Reply:
Done